source/common/rate_limiter_impl.cc

```diff
@@ -6,6 +6,50 @@
 
 namespace Nighthawk {
 
+BurstingRateLimiter::BurstingRateLimiter(RateLimiterPtr&& rate_limiter, const uint64_t burst_size)
+    : rate_limiter_(std::move(rate_limiter)), burst_size_(burst_size) {
+  ASSERT(burst_size_ > 0);
+}
+
+bool BurstingRateLimiter::tryAcquireOne() {
+  // Callers may not be able to actually make use of a successfull acquisition, and
+  // call releaseOne() to indicate so. We remember state here so we can use that later when
+  // that happens to restore it.
+  previously_releasing_ = releasing_;
+
+  if (releasing_) {
+    // When releasing_ we drain the accumulated total until there's nothing left. We'll transit to
+    // accumulating mode after that.
+    accumulated_--;
+    releasing_ = accumulated_ > 0;
+    return true;
+  } else {
+    // We are greedy here, which should help with timely releases of bursts when
+    // the system is lagging behind.
+    while (rate_limiter_->tryAcquireOne()) {
+      accumulated_++;
+      if ((accumulated_ % burst_size_) == 0) {
+        releasing_ = true;
+        // We have accumulated the burst size. Enter release mode and recurse.
+        return tryAcquireOne();
+      }
+    }
+  }
+
+  previously_releasing_ = absl::nullopt;
+  return false;
+}
+
+void BurstingRateLimiter::releaseOne() {
+  ASSERT(accumulated_ < burst_size_);
+  ASSERT(previously_releasing_ != absl::nullopt && previously_releasing_ == true);
+  // The caller wasn't able to put its earlier successfull acquisition to good use, so we restore
```

Comment:
Oh, this is actually different to what I thought you were doing. I thought you were doing:
* During acquisition, keep increasing accumulated until % burst_size == 0.
* During release, keep increasing accumulated until % burst_size ==0.

If you ++ and --, won't you run the risk of getting non-bursty behavior when released tokens get immediately reacquired and a thread spends a long period without requesting a burst?